src/scripts/main.js

```diff
@@ -1,3 +1,203 @@
 'use strict';
 
-// write your code here
+const messageStart = document.querySelector('.message-start');
+const messageWin = document.querySelector('.message-win');
+const messageLose = document.querySelector('.message-lose');
+
+const gameField = document.querySelector('.game-field');
+const gameFieldWidth = 4;
+
+const buttonStart = document.querySelector('.start');
+const gameScore = document.querySelector('.game-score');
+
+const winScore = 2048;
+let score = 0;
+
+let gameBoard = boardInit();
+
+function boardInit() {
+  return Array.from({ length: gameFieldWidth },
+    () => Array(gameFieldWidth).fill(0));
+}
+
+function startGame() {
+  messageStart.classList.add('hidden');
+  buttonStart.classList.remove('start');
+  buttonStart.classList.add('restart');
+  buttonStart.innerText = 'Restart';
+
+  resetGameBoard();
+  fillRandomCell();
+  fillRandomCell();
+  fillBoard();
+
+  document.addEventListener('keyup', movesController);
+}
+
+function restartGame() {
+  resetGameBoard();
+  gameScore.innerText = 0;
+  score = 0;
+  startGame();
+  messageWin.classList.add('hidden');
+  messageLose.classList.add('hidden');
+}
+
+function resetGameBoard() {
+  gameBoard = boardInit();
+}
+
+function handleWin() {
+  const userWon = gameBoard.flat().includes(winScore);
+
+  if (userWon) {
+    messageWin.classList.remove('hidden');
+    buttonStart.classList.remove('restart');
+    buttonStart.classList.add('start');
+    buttonStart.innerText = 'Start';
+
+    resetGameBoard();
+    document.removeEventListener('keyup', movesController);
+  }
+}
+
+function handleLose() {
+  const hasEmptyCell = gameBoard.flat().some((cell) => cell === 0);
+
+  if (!hasEmptyCell) {
+    messageLose.classList.remove('hidden');
+    messageWin.classList.add('hidden');
+    document.removeEventListener('keyup', movesController);
+  }
+}
+
+function randomNumber() {
+  return Math.floor(Math.random() * gameFieldWidth);
+}
+
+function fillRandomCell() {
+  const [row, col] = [randomNumber(), randomNumber()];
+
+  if (!gameBoard[row][col]) {
+    gameBoard[row][col] = Math.random() > 0.9 ? 4 : 2;
+    handleLose();
+  } else {
+    fillRandomCell();
+  }
+}
+
+function fillBoard() {
+  gameBoard.forEach((row, rowCount) => {
+    row.forEach((_, cellCount) => {
+      const tile = gameField.rows[rowCount].cells[cellCount];
+      const tileValue = gameBoard[rowCount][cellCount];
+
+      tile.classList = 'field-cell';
+      tile.innerText = tileValue;
+
+      if (tileValue > 0) {
+        tile.classList.add('field-cell--' + tileValue);
+      } else {
+        tile.innerText = '';
+      }
+    });
+  });
+}
+
+function duplicateGameBoard() {
+  return gameBoard.map((row) => [...row]);
```

Comment:
Instead of using 'map', you can use 'slice()' method to create a shallow copy of the gameBoard. Replace 'return gameBoard.map((row) => [...row]);' with 'return gameBoard.map((row) => row.slice());'